support removal of event handlers from SharedIndexInformers

mandelsoft opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:
To be able to implement controllers that are dynamically deciding
on which resources to watch, it is required to get rid of
dedicated watches and event handlers again. This requires the
possibility to remove event handlers from `SharedIndexInformer`s again.
Stopping an informer is not sufficient, because there might
be multiple controllers in a controller manager that independently
decide which resources to watch.

Unfortunately the ResourceEventHandler interface encourages to use
value objects for handlers (like the ResourceEventHandlerFuncs
struct, that uses value receivers to implement the interface).
Go does not support comparison of function pointers and therefore
the comparison of such structs is not possible, also. To be able
to remove all kinds of handlers and to solve the problem of
multi-registrations of handlers a registration handle is introduced.
It is returned when adding a handler and can later be used to remove
the registration again. This handle directly stores the created
listener to simplify the deletion.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
Remark: If as the result of a handler removal a complete informer
should be disabled it is higly recommended to incorporate
pull request #98653, which fixes a race condition when stopping
watches for an informer using the stop channel.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mandelsoft committed:
support removal of event handlers from SharedIndexInformers

To be able to implement controllers that are dynamically deciding
on which resources to watch, it is required to get rid of
dedicated watches and event handlers again. This requires the
possibility to remove event handlers from SharedIndexInformers again.
Stopping an informer is not sufficient, because there might
be multiple controllers in a controller manager that independently
decide which resources to watch.

Unfortunately the ResourceEventHandler interface encourages to use
value objects for handlers (like the ResourceEventHandlerFuncs
struct, that uses value receivers to implement the interface).
Go does not support comparison of function pointers and therefore
the comparison of such structs is not possible, also. To be able
to remove all kinds of handlers and to solve the problem of
multi-registrations of handlers a registration handle is introduced.
It is returned when adding a handler and can later be used to remove
the registration again. This handle directly stores the created
listener to simplify the deletion.

fedebongio commented:
/assign @kevindelgado 
/triage accepted

kevindelgado commented:
/assign @DirectXMan12

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -666,6 +745,35 @@ func (p *sharedProcessor) resyncCheckPeriodChanged(resyncCheckPeriod time.Durati
+	for i, l := range p.listeners {
+		if reflect.ValueOf(l.handler).Type().Comparable() && l.handler == handler {
+			listener = l
+			p.listeners = append(p.listeners[:i], p.listeners[i+1:]...)
```
I think you will get a `slice bounds out of range` panic here because you are modifying the size of slice as you iterate over it (e.g if the handler to be removed exists at any spot in the slice and at the end).

mandelsoft replied:
You are right, if there would be multiple matches. To be safe I'll change this to an index based loop

mandelsoft replied:
There are even more issues, if there are multiple registrations for the same handler. All the listeners must be stopped, not only the first/last one.
I fixed this accordingly.

kevindelgado commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -666,6 +745,35 @@ func (p *sharedProcessor) resyncCheckPeriodChanged(resyncCheckPeriod time.Durati
+		return nil
+	}
+	var listener *processorListener
+	for i, l := range p.listeners {
```
Note that you have made a decision to remove all occurrences of the handler, in the event that a handler is added multiple times. That's probably fine (and there's not really a good reason to ever add a handler more than once), but I would explicitly call out this decision in a comment somewhere, in case users are expecting different behavior.

mandelsoft replied:
I add a note to the method definition and interface comment

kevindelgado reviewed (commented):
Hi Uwe, great minds think alike (or fools seldom differ)! I submitted an almost identical PR a bit before yours over here at https://github.com/kubernetes/kubernetes/pull/98536. We can go ahead with yours though since I think it might be a bit more complete than mine, but I’ve left a few comments/questions on some of the parts where we differ.

I was adding this change as part of some other work I had been doing related to giving finer grained control over the lifecycle of SharedIndexInformers at https://github.com/kubernetes/kubernetes/pull/97214 I’d love for you to take a look or give it a review that if you have time and are interested.

I’m tagging Yu Liao from sig-api-machinery on this because he has been helping review my other work in this area (and I’m not an org member so I don’t have approval access).

kevindelgado commented:
/unassign @DirectXMan12 
/assign @yliaog

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +571,83 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+// It offers the removing of handlers and some state info useful
+// to use after handler removal to decide whether an informer
+// is still required.
+type ExtendedSharedIndexInformer interface {
```
I did not know whether the informer interface is used by other implementations, also. In such a case extending the interface would break all those other implementations.

This was the reason I decided not to extend the existing interface, but introduce ab extended one.

If you don't see a compatibility problem here, I surely can extend the existing interface, instead,

kevindelgado replied:
I know in kubernetes/kubernetes there are no other implementations, not sure about outside of k/k. I'll leave it @yliaog or someone from sig-api-machinery to decide.

yliaog replied:
it is a breaking change if extending the SharedIndexInformer Interface directly. For breaking change, a few options: 1) send an email to kubernetes-sig-api-machinery@google.com, or kubernetes-dev@google.com 2) discuss it in biweekly sig-api-machinery meeting. 3) add to release notes about the breaking change

I think it is cleaner to extend the SharedIndexInformer directly, but it's better to check the community to ensure it won't break people unintentionally.

mandelsoft replied:
I've sent I mail to  kubernetes-sig-api-machinery@google.com referring to this PR

lavalamp replied:
I'll put it on the agenda; I can see either breaking compat or adding a "EventHandlerRemovable" as above, but I can't see making this "ExtendedSharedIndexInformer" interface; what will we do when we add another method, make a "ExtendedExtendedSharedIndexInformer"?

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +571,83 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	SharedIndexInformer
+	EventHandlerRemovable
+	HasEventHandlers() bool
+	IsStopped() bool
```
It gives you more control managing informers besides factories, once you have bare informers in your hands. and you want to explicitly create/delete watches and/or add/remove handlers again.

kevindelgado replied:
makes sense

mandelsoft replied:
The creation and deletion of watches is not part of the PR, it could completely be done on top. The scenario for the usage of such a feature are controllers that are controlled by other resources. Just an example, you want to provide indices of arbitrary kubernetes resources, based on a dedicated CRD. Once an index is created for a resource, a watch has to be established (or if already existent, only an additional handler might be configured). If the index is deleted again, the handler and, if the watch is not used by other controllers in the same controller manager, the watch should be removed again. Similar actions would be required, if a controller is potentially responsible for multiple clusters, depending on certain conditions provided by other resources.

The deletion of watches is possible just by a dedicated cancel context (on top) used to initiate the watch and the informer. This works fine (beside the problem described in  #98653) and is minimal invasive. So, all this is potentially possible just by adding the feature of being able to remove handlers again.

If a traditional InformerFactory is used this is for sure not possible without extending the factory, too. But this PR is the precondition to propagate such a feature up the abstraction levels and focuses on the informer itself.

yliaog replied:
thanks for the explanations. i'm not sure i fully understand the use case, two quick questions:
1. do you still need a separate watch if informer is used? put it another way, wouldn't an informer for a resource sufficient, hence no need to use both an informer and a watch on the same resource?
2. since this PR is about removing event handlers, probably it's better to have a separate PR for supporting watches creation/deletion if needed. what do you think?

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +571,83 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+
+// HasEventHandlers reports whether the informer still has registered
+// event handlers
+func (s *sharedIndexInformer) HasEventHandlers() bool {
```
For the functionality o managing informers this would not be necessary, but you are right, for debugging it might be useful.
I changed the name and semantics accordingly

kevindelgado commented:
lgtm % a final decision on extending/addding to the shared informer interface. Leaving that up to a reviewer sig-api-machinery.

yliaog commented:
/ok-to-test

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +571,84 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	}
+
+	if !reflect.ValueOf(handler).Type().Comparable() {
+		return fmt.Errorf("uncomparable handler")
```
better to add more info, something like "Uncomparable handler %v was not removed"

mandelsoft replied:
you are right, just changed

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -666,6 +746,41 @@ func (p *sharedProcessor) resyncCheckPeriodChanged(resyncCheckPeriod time.Durati
+	defer p.listenersLock.Unlock()
+
+	listeners := p.removeListenerLockedFor(handler)
+	if p.listenersStarted && len(listeners) > 0 {
```
len(listeners) >0 is likely not necessary

mandelsoft replied:
just changed

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,93 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	handler2 := &ResourceEventHandlerFuncs{}
+	informer.AddEventHandler(handler2)
+
+	if informer.EventHandlerCount() == 0 {
```
why not checking "informer.EventHandlerCount() != 2" to be precise

mandelsoft replied:
yes, the test was originally intended for the bool HasEventHandler function before changing it according to a change request for this PR. I'll change this.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,93 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	informer.AddEventHandler(handler2)
+
+	if informer.EventHandlerCount() == 0 {
+		t.Errorf("informer pretends not have a registered handler")
```
the error message needs to be changed too.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,93 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	}
+
+	if err := informer.RemoveEventHandler(handler2); err != nil {
+		t.Errorf("removing on pointer handler failed: %s", err)
```
s/removing on pointer handler/removing pointer handler/ ?

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,93 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	if err := informer.RemoveEventHandler(handler2); err != nil {
+		t.Errorf("removing on pointer handler failed: %s", err)
+	}
+	if informer.EventHandlerCount() == 0 {
```
what about checking "informer.EventHandlerCount() != 1" ?

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,93 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	if err := informer.RemoveEventHandler(handler1); err == nil {
+		t.Errorf("removing on value handler does not fail")
+	} else {
+		if err.Error() != "uncomparable handler" {
```
this message has been changed, so checking with strings.Contains(...)?

mandelsoft commented:
@yliaog  With the actual state for the pure watch removal no further changes are required. The additional `Is...` methods just completes the `SharedIndexInformer` interface concerning the handler registration allowing to work with bare informers. 
For sure, there is no additional watch required besides the one behind the informer. But during dynamic handler registration/deregistration it is required to know whether the informer has still other handlers and whether it has already been started to decide whether it needs to be stopped to ensure the watch is canceled. This way these operations just complete the lifecycle api of a shared informer.

The optional extension of the upper API layer, the informer factories, to support cancelling of dedicated informers should certainly be done separately. But the factory so far does not know about handler registrations, which are directly performed on the bare informers. Therefore it might be difficult to completely hide the informer management in the factories, besides the feature of stopping an informer (and removing it from the factory again).
This one just provides a cohesive and clear interface for a shared informer enabling building things on top.

mandelsoft commented:
I've added the new methods to the `SharedInformer` interface and removed the additional ones, again.
BTW: should I extend the `SharedIndexInformer` or the `SharedInformer` interface. Basically it seems to be an extension of the interface for the plain informer. The index informer just adds the indexer functionality.

mandelsoft commented:
/retest

mandelsoft commented:
/retest

mandelsoft commented:
Hi @yliaog, can I still do something?

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +592,55 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+}
+
+// RemoveEventHandler tries to remove a formerly added event handler again.
+// Only event handlers that are (go) comparable can be removed again.
```
ditto

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +592,55 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	}
+
+	if !reflect.ValueOf(handler).Type().Comparable() {
+		return fmt.Errorf("Uncomparable handler %v was not removed", handler)
```
s/was/is/

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -666,6 +738,41 @@ func (p *sharedProcessor) resyncCheckPeriodChanged(resyncCheckPeriod time.Durati
+func (p *sharedProcessor) removeListenerLockedFor(handler ResourceEventHandler) []*processorListener {
+	if !reflect.ValueOf(handler).Type().Comparable() {
+		return nil
+	}
```
this has been checked at line 630, does it need to be checked again?

mandelsoft replied:
not really, for the actual call stack,  but the method might later on be called by other callers in a different context, and for the sake of defensive programming, I've added this to provide a safe function. 

But if you want I'll remove the check again.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,139 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+		t.Errorf("removing of first pointer handler failed: %s", err)
+	}
+	if informer.EventHandlerCount() != 1 {
+		t.Errorf("after removing handler informer pretends to have %d registered handler(s), instead of 1", informer.EventHandlerCount())
```
remove 'pretends'?

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,139 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+		t.Errorf("removing of second pointer handler failed: %s", err)
+	}
+	if informer.EventHandlerCount() != 0 {
+		t.Errorf("informer pretends to still have registered handlers after removing both handlers")
```
ditto, and also in the tests below

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,139 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	}
+
+	if !informer.IsStarted() {
+		t.Errorf("informer does not report to be started although handling events")
```
ditto

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,139 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+		return
+	}
+	if informer.IsStopped() {
+		t.Errorf("informer reports to be stopped although stop channel not closed")
```
ditto

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,139 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	close(stop)
+	fmt.Println("sleeping")
+	time.Sleep(1 * time.Second)
+
```
add a case for: removing event handler from stopped informer?

kevindelgado commented:
Friendly ping @mandelsoft

mandelsoft commented:
Sorry I was busy this week, I'll continue over the weekend

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -614,21 +614,21 @@ func (s *sharedIndexInformer) EventHandlerCount() int {
 	if s.stopped {
-		klog.V(2).Infof("Handler %v was not removed from shared informer because it has stopped already", handler)
-		return nil
+		klog.V(2).Infof("Handler %v is not removed from shared informer because it has stopped already", handler)
```
the logging here can be removed, since it is returned in the error below.

kevindelgado commented:
Is this just waiting on another review from @yliaog ?

yliaog commented:
/lgtm

please squash the commits

kevindelgado commented:
friendly ping @mandelsoft (can you squash the commits)

mandelsoft commented:
@yliaog  I've rebased and squashed

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +592,54 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	defer s.startedLock.Unlock()
+
+	if s.stopped {
+		return fmt.Errorf("Handler %v is not removed from shared informer because it has stopped already", handler)
```
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

error string is better to start with lowercase

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +592,54 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	}
+
+	if !reflect.ValueOf(handler).Type().Comparable() {
+		return fmt.Errorf("Uncomparable handler %v is not removed", handler)
```
ditto

yliaog reviewed (commented):
some nit comments, otherwise looks good to me. Thanks.

yliaog commented:
/lgtm
/approve

yliaog commented:
/retest

mandelsoft commented:
/restest

mandelsoft commented:
/retest

mandelsoft commented:
/retest

mandelsoft commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -150,6 +152,15 @@ type SharedInformer interface {
 	// be competing load and scheduling noise.
 	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration)
+	// RemoveEventHandler removes a formerly added event handler.
+	// It is possible to remove only event handlers that are (go) comparable,
```
This is a strange restriction for a method signature that takes an interface. It's also disconcerting since most ResourceEventHandler calls use non-pointer ResourceEventHandlerFuncs, and the error when trying to remove is a runtime error, not a compile error.

Either returning an handle that could be passed to RemoveEventHandler to remove the handler (regardless of whether the original handler was comparable), or passing in something that could be used to explicitly remove the handler (stopChannel that can be closed, etc) might be better.

mandelsoft replied:
The I would prefer to return a handle. But anyway this is much more invasive change, because the internal 
structure has to be changed. I'll try to prepare a proposal the next days.

mandelsoft replied:
May be we could wrap the uncomparable handler into a handle implementing the ResourceEventHandler interface. So, if the handler is comparable, we use the handler as handle, and if not it is wrapped. 
The Add methods hist return the resulting effective handler, that can always be removed.

Or do you pefer a more explicit Handle interface. This could avoid the runtime exception at all.

mandelsoft replied:
There are already other implementations for a `SharedIndexInformer` interface wrapping a base informer.
Those will break if I change the signature of an existing method.
Is it ok to change the signature (and adapting the implementations in the kubernetes project)?

mandelsoft replied:
@liggitt I've added a commit introducing an explicit handle. I could stick to the handler interface internally. Therefore the change is less invasive than expected. But the signature change broke a wrapper of a shared informer, Therefore I had to change something in the replication controller, also.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +192,28 @@ type SharedInformer interface {
+// interface can be implemented by non-go-comparable handlers, which
+// could not be removed from a list anymore.
+type ResourceEventHandlerHandle struct {
+	ResourceEventHandler
```
is there a reason we need to make this public and inline it? making it private and not inlined makes the handle more opaque outside this package, which seems good

mandelsoft replied:
I did the inlining to avoid changes to the lower coding layers. But you are right, it should not be exposed to the caller, because then he could change it.

There are two possibilities to keep the lower layers:
- make the field private and implement the handler interface for the handle (that's what I've done with the next commit)
- add another invisible indirection layer by internally registering a wrapped handler and return a handle referring to the wrapped handler to be able to remove it again

What do you prefer?

liggitt replied:
yeah, if the handle just contained the listener pointer as a private field, it wouldn't need to implement or expose anything, and could be an opaque pure handle

mandelsoft replied:
done

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -474,14 +508,15 @@ func determineResyncPeriod(desired, check time.Duration) time.Duration {
 	if s.stopped {
 		klog.V(2).Infof("Handler %v was not added to shared informer because it has stopped already", handler)
-		return
+		return nil
```
if we have the possibility of returning nil, this should probably return `handle, bool` to make it clear we don't always return handles

mandelsoft replied:
I added an error return here.

mandelsoft replied:
let's continue this discussion with the review below

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +605,50 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	return nil
 }
 
+// IsStarted reports whether the informer has already been started
+func (s *sharedIndexInformer) IsStarted() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.started
+}
+
+// IsStopped reports whether the informer has already been stopped
+func (s *sharedIndexInformer) IsStopped() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.stopped
+}
+
+// EventHandlerCount reports whether the informer still has registered
+// event handlers
+func (s *sharedIndexInformer) EventHandlerCount() int {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return len(s.processor.listeners)
+}
```
They are intended for a informer management layer on top, that manages the handler and watches.
For sure, this must be synchronized again. You are right that the naked usage of those methods bear the potential for race conditions. But without them it would not be possible to offer higher layers outside the cache package.

liggitt replied:
> it's not clear to me how they are going to be used (i.e. the informer management layer on top). suggest to make it private for this PR, so that it is focused on supporting removal of event handler only.

agreed

lavalamp replied:
I'm just doing a pre-review skim of this PR, this comment (make these three private) doesn't seem to have been applied.

I'm not sure what "higher layers" might need to do but we can enable it specifically when we figure it out.

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +605,50 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	// 3. unblock
+	s.blockDeltas.Lock()
+	defer s.blockDeltas.Unlock()
+	s.processor.removeListenerFor(handle)
```
Good idea, because now single registrations and not handlers are removed the complete handling of removing multiple listeners get obsolete. I'll change this later. Now I've to leave  for two hours.

Please give me an advise, which handle model (see above) you prefer. I've already implemented the second solution, but not yet pushed it.

liggitt replied:
> make the field private and implement the handler interface for the handle (that's what I've done with the next commit)

I don't think the handle we return should implement the handler interface.

> add another invisible indirection layer by internally registering a wrapped handler and return a handle referring to the wrapped handler to be able to remove it again

something along those lines sounds good... I think the listener we already created internally can serve as the wrapped handler, right?

mandelsoft replied:
Yepp, directly using the listener as proposed by @liggitt instead of a wrapper just implies the second behaviour. Now the handle is just an anonymous handle for the outside world and directly refers to the created listener already wrapping the handler.

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +192,28 @@ type SharedInformer interface {
 	// offloaded.
 	SetWatchErrorHandler(handler WatchErrorHandler) error
+
+	// EventHandlerCount return the number of actually registered
+	// event handlers.
+	EventHandlerCount() int
+
+	// IsStopped reports whether the informer has already been stopped.
+	// Adding event handlers to already stopped informers is not possible.
+	IsStopped() bool
+
+	// IsStarted reports whether the informer has already been started
+	IsStarted() bool
```
The problem is, that they should be used outside the package, in a higher informer management layer, therefore they cannot be private. They will not only be used for testing (see my comment above)

lavalamp replied:
We need to refactor the interface until the things people need to do can be done race-free. E.g. instead of calling IsStopped() before adding a handler, just add the handler and it gives you an error if the informer has been stopped.

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -508,13 +518,13 @@ func determineResyncPeriod(desired, check time.Duration) time.Duration {
 	if s.stopped {
 		klog.V(2).Infof("Handler %v was not added to shared informer because it has stopped already", handler)
-		return nil
+		return nil, fmt.Errorf("handler %v was not added to shared informer because it has stopped already", handler)
```
do you need an error? if it is added successfully, it is not going to be nil. nil means there is an error.

liggitt replied:
either an error or a bool second return value makes it clear that the function can fail

without that, we're asking for people to not handle the nil case correctly

yliaog replied:
that's right. i think bool second return is better.

mandelsoft replied:
Hmm, I think the Add and Remove methods should behave similar in case of used for a stopped handler. That is the same reason. Either both return an additional bool or an error. Just for symmetry.

In case of an error for the Add methods we also could get rid of the log message as requested for the remove method.

yliaog replied:
*all* the currently existing calls of AddEventHandler do *not* check the return (since there is no return before this code change), when it is upgraded to the version with this code change, and there is an error, the error will be silently dropped (no log lines printed out).

mandelsoft replied:
Ok, but then we should use in both cases a bool return

yliaog replied:
I think i agree with you considering that there might be future error cases. For the Add method, since there is currently a log line, i think it is better to keep it to avoid surprises to the existing code users who does not handle the return (since there is no return before the change).

@liggitt what do you think about it?

liggitt replied:
returning error lets you communicate details about the error, that seems reasonable to me

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -503,11 +538,11 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 
-	listener := newProcessListener(handler, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize)
+	listener := newProcessListener(handle, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize)
 
```
definitely

mandelsoft commented:
/retest

mandelsoft commented:
/retest

mandelsoft commented:
@yliaog, how to proceed? I've switched to the bool return as requested (to both, Add and Remove). But because of the arguments mentioned above I would prefer the error return.

kevindelgado commented:
friendly ping @yliaog @liggitt

yliaog commented:
@mandelsoft please use error instead of bool as discussed. thanks.

kevindelgado commented:
friendly ping @mandelsoft

yliaog commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +357,263 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	if err == nil {
+		t.Errorf("stopped informer did not reject add handler")
+		return
+	} else {
```
'else' can be removed

mandelsoft commented:
> it's not clear to me how they are going to be used (i.e. the informer management layer on top). suggest to make it private for this PR, so that it is focused on supporting removal of event handler only.
>
> if needed, have another PR to tackle the support for the informer management layer on top.

@yliaog, The reason for adding the remove functionality is, that it should be possible to dynamically manage informers according to the need of a controller. Therefore it might happen, that, in a controller manager hosting multiple (such) controllers, handlers are removed and without a handler an informer could be stopped. So, when adding a handler again, it might happen that the informer is already stopped (by the stop channel). So it would be useful to have a formal way to access the state of an informer by a small management layer on-top of the informers, to be able to create/stop/destroy informers according to actually desired set of handlers. The informer already tracks its state, so it would be useful to have the possibility to access this state without the need of maintaining duplicate state at the management level. 

I don't see why these additional informational methods should be harmful. Yes, using them to create/start/stop informers will require an external synchronization, but they could avoid the management of duplicate state.

yliaog reviewed (commented):
@mandelsoft have you checked out the PR from @kevindelgado on informer lifecycle improvement, https://github.com/kubernetes/kubernetes/pull/97214#issuecomment-810609141, does that meet your use case?

i'm fine with adding these additional informational methods if they are really necessary, probably add comments around the need for external synchronization.

mandelsoft commented:
I've added an explanation for the informational methods to the description of the `SharedInformer` interface (and the log message is back again for the Add method)

yliaog commented:
thanks for the pr. it looks good to me. i'll leave it to @liggitt to take another look. please squash the commits.

/lgtm

mandelsoft commented:
commits squashed and rebased

mandelsoft commented:
/retest

mandelsoft commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +617,54 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+
+	if handle.listener == nil {
+		return fmt.Errorf("registration already removed")
+	}
```
move this check before the lock. I also think this should probably return nil, since the desired state (the listener associated with the handle is removed) still holds if there is no associated listener (it makes calling remove twice on the same handle an idempotent operation)

mandelsoft replied:
I've additionally added a method to the handle to check the handle state

kevindelgado commented:
friendly ping @liggitt

kevindelgado commented:
friendly ping @liggitt

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -130,11 +131,24 @@ import (
+// are intended to be used to manage informers in any upper informer
+// management layer for creating and destroying informers on-the fly when
+// adding or removing handlers.
+// Beware of race conditions: Such a layer must hide the basic informer
```
This is too big of a caveat IMO, we have to choose a different paradigm. Perhaps the higher layer could e.g. register callbacks that can be called by the shared informer at the appropriate times.

mandelsoft replied:
I've removed those methods again causing the potential for races. `IsStopped` keeps stopped forever once true is reported. Therefore we should be able to keep it.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -149,7 +163,20 @@ type SharedInformer interface {
+	// RemoveEventHandlerByHandle removes a formerly added event handler given by
+	// its registration handle.
+	// If, for some reason, the same handler has been added multiple
+	// times, only the registration for the given registration handle
```
shouldn't each time get a unique handle?

mandelsoft replied:
The handle is unique for a registration, not for the handler, you can still add a handler multiple times, and remove the registrations one by one. It would be dangerous to enforce the removal of a handler registered multiple times, if only one registration is cancelled. 

The reason for introducing a handler was that handlers might not be comparable. Because of this it is also no possible to provide a handle unique for a handler.

lavalamp replied:
I'm having trouble parsing what you mean by "registration" and "handler".

Suppose I call this N times, passing ResourceEventHandler `h` each time. I would then expect that when a matching event arrives, `h` is notified N times. I would further expect that each call to AddEventHandlerWithResyncPeriod produced a different ResourceEventHandlerHandle and I would have to call remove for each handler handle to stop getting notifications.

I'd further expect that if I add `h` and someone else adds handler `g` which happens to match the same events, that when I remove my ResourceEventHandlerHandle, `g` would not be affected.

If it doesn't work like this, please make it work like this.

Please update the comment to make this clearer.

mandelsoft replied:
Exactly this is how is works. With every add call a new new registration handle for the handler is created, which results in an additional listener, which uses the same handler. Therefore it is called n times, if it is added n times. The same is true for the deletion; the remove call for a handler just removes the listener associated with the handle wich refers to the originally given handler. Therefore there will be a decrementing amount of calls with every remove operation for an not already removed handle originally created for the same handler.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +207,35 @@ type SharedInformer interface {
+// meaning that the handler registered with this handle is
+// still registered.
+func (h *ResourceEventHandlerHandle) IsActive() bool {
+	return h.listener != nil
```
This looks likely to be racy. Do we really need this method?

mandelsoft replied:
removed again

lavalamp replied:
It appears not removed? :)

I expected implementing this to need to make a call to a SharedInformer, or for this to be a method of a SharedInformer which takes in a ResourceEventHandlerHandle. But it is best if it just doesn't exist at all (and it is safe to remove ResourceEventHandlerHandle multiple times).

mandelsoft replied:
yup, sorry, you are right. I made it local now to be still usable for the test coding. Additionally the modification of its field (listener) is now inside the synchronized code for the shared informer.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +207,35 @@ type SharedInformer interface {
+
+	// EventHandlerCount return the number of actually registered
+	// event handlers.
+	EventHandlerCount() int
```
Why is this needed?

mandelsoft replied:
removed again

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -149,7 +163,20 @@ type SharedInformer interface {
+	// Calling Remove on an already removed handle returns no error
+	// because the handler is finally (still) removed after calling this
+	// method.
+	RemoveEventHandlerByHandle(handle *ResourceEventHandlerHandle) error
```
ResourceEventHandlerHandle contains a pointer, so you don't need to take it as a pointer, no?

mandelsoft replied:
You are right, a value object would work and is even comparable, But returning a value object would require to return a handle even if the operation was failing, because it is not possible to return a `nil`.
This would require to figure out whether it is a valid handle or not.

For a user of the method, typically the concrete type is irrelevant, the user just uses the return value, regardless whether it is a pointer or not. So, for the sake of having a general `nil` I would prefer a pointer.

lavalamp replied:
This isn't super important to me, but I don't see why it's important to be able to return a nil. Users of this library shouldn't act on any handle if it is gotten with an error, and I don't think they should need to compare to nil. Also since this has an internal pointer, the zero value should act more or less like nil.

mandelsoft replied:
The listener stored in the registration handle is used to unregister the handle again. Afterwards it is reset to nil (synchronized under the informer lock). If a value object would be used for the handle, it would not possible anymore to reset this fielld and the handler would always be valid for unregistration. This was the reason for the `IsActive` method I removed on your request. Nevertheless the listener field is still used for this purpose.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -130,11 +131,24 @@ import (
-	AddEventHandler(handler ResourceEventHandler)
+	// It returns a handle for the handler that can be used to remove
+	// the handler again.
+	AddEventHandler(handler ResourceEventHandler) (*ResourceEventHandlerHandle, error)
```
I'm not sure it's legit to start returning an error; that will leave preexisting code likely still compiling but not handling the error.

Instead I think it is safer to add a new method ("AddRemovableEventHandler"). Same comment for below method.

All of these interface changes break implementors of the interface; we should either know that that is very rare, or we should make SharedInformerV2 interface.

mandelsoft replied:
This method might have been failed even before (for stopped informers). But there was no way for a caller to react on this, there was just a log message.

With this change this behavior for actual callers in unchanged and the coding does not get worse. But now the caller has the change to figure out that something went wrong and what went wrong. He is still free to decide to ignore it. 

May be there are warnings in some environments now, that the error return is ignored, but this hint gives you the chance to remember that here the coding could be improved.

Because of the interface change I had started a discussion in the mailing list months ago, and it it was approved.

lavalamp replied:
I found that email thread, it looks like past!me was expecting an additional method in this same interface. But it looks like everyone was in favor so I won't insist.

kevindelgado commented:
friendly ping @mandelsoft

mandelsoft committed:
remove informational informer methods again

atoato88 commented:
gentle ping @mandelsoft

kevindelgado commented:
@lavalamp I think this ready for another review from you if you have time.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +197,29 @@ type SharedInformer interface {
+// This indirection is required, because the ResourceEventHandler
+// interface can be implemented by non-go-comparable handlers, which
+// could not be removed from a list anymore.
+type ResourceEventHandlerHandle struct {
```
Can we call this a `ResourceEventRegistration`? "HandlerHandle" is really awkward to talk about.

mandelsoft replied:
Ok, but it is nor a registration of a resource event. Then we should name it `ResourceEventHandlerRegistration`, because it is a registration of a (resource event) handler.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +197,29 @@ type SharedInformer interface {
+// interface can be implemented by non-go-comparable handlers, which
+// could not be removed from a list anymore.
+type ResourceEventHandlerHandle struct {
+	listener *processorListener
```
I expected to find a key to a map here :) (I haven't read the implementation)

mandelsoft replied:
I made it constant now and switched to a value return for the handle. 

The problem basically was something different. If a handle was removed from an informer it was NOT registered for,
it got corrupted. Now you can remove a handle from any informer without any harm.
The operation is only executed, if the handle is still active for the informer it is passed to remove it.

lavalamp replied:
IMO, no, the removal should always succeed (i.e., be safe to call twice). This makes it very convenient if a user has multiple shutdown paths.

mandelsoft committed:
rename handle to registration

mandelsoft commented:
/retest

lavalamp reviewed (commented):
Sorry for delay. I have to argue with the registration type some more :)

mandelsoft committed:
apply desired changes for handler registration

mandelsoft committed:
tests for invalid registration removals

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +608,40 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+
+	if handle.listener == nil {
```
Remove this, we don't set these to nil anymore so this won't optimize anything, right?

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -584,6 +657,19 @@ type sharedProcessor struct {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	for i := 0; i < len(p.listeners); i++ {
```
Is there any reason not to switch the lists to maps/sets? e.g. make the type `map[*processorListener]struct{}`. Then this check and the removal are easy.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -600,6 +686,37 @@ func (p *sharedProcessor) addListenerLocked(listener *processorListener) {
+			found = true
+		}
+	}
+	for i := 0; i < len(p.syncingListeners); i++ {
```
You should only need to do this second search if found is true, right?

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -600,6 +686,37 @@ func (p *sharedProcessor) addListenerLocked(listener *processorListener) {
+		if l == listener {
+			p.listeners = append(p.listeners[:i], p.listeners[i+1:]...)
+			i--
+			found = true
```
You can break here unless the listener can get in the loop twice? (another reason to switch to maps)

mandelsoft replied:
break added and i-- removed. Basically the former coding was more defensive.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -600,6 +686,37 @@ func (p *sharedProcessor) addListenerLocked(listener *processorListener) {
+	}
+}
+
+func (p *sharedProcessor) removeListenerLocked(listener *processorListener) bool {
```
Since this is only called one place, I'm not sure it really helps readability to split it out into a separate function like this, I personally would not.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -474,13 +511,13 @@ func determineResyncPeriod(desired, check time.Duration) time.Duration {
 	if s.stopped {
-		klog.V(2).Infof("Handler %v was not added to shared informer because it has stopped already", handler)
-		return
+		klog.V(2).Infof("Handler %v is not added to shared informer because it has stopped already", handler)
```
I think we should remove the log line now that we're returning an error. It's a little rude for libraries to log internally anyway, it limits choices of the consumer.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -149,7 +153,20 @@ type SharedInformer interface {
+	// Calling Remove on an already removed handle returns no error
+	// because the handler is finally (still) removed after calling this
+	// method.
+	RemoveEventHandlerByRegistration(handle ResourceEventHandlerRegistration) error
```
I'd leave off the "ByRegistration" part -- you can tell what you need by the function signature.

mandelsoft replied:
suffix removed

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +376,377 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	}
+
+	if eventHandlerCount(informer) != 2 {
+		t.Errorf("informer has %d registered handler(s), instead of 1", eventHandlerCount(informer))
```
```suggestion
		t.Errorf("informer has %d registered handler(s), instead of 2", eventHandlerCount(informer))
```

mandelsoft replied:
fixed

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +376,377 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+	close(stop)
+	fmt.Println("sleeping")
+	time.Sleep(1 * time.Second)
+	if !informer.IsStopped() {
```
Personally I would just wait in a loop until the condition is true, sleep 50ms between each check. Waiting a second and then failing might be flaky.

mandelsoft replied:
I've generalized the `ok` methods with for a polled check with timeout to use this here, also.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +376,377 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+		return
+	}
+	err = informer.RemoveEventHandlerByRegistration(handle)
+	if err == nil {
```
As stated above, I don't think this should be an error.

mandelsoft replied:
error case removed. But I left the error return to be prepared for further checks. For example whether the handle is valid for the actual informer.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
```diff
@@ -357,3 +376,377 @@ func TestSharedInformerErrorHandling(t *testing.T) {
+		t.Errorf("removing handler from stopped informer yield unexpected error: %s", err)
+		return
+	}
+}
```
Missing test: add and remove handlers while an informer is running and processing items, to detect any locking problems.

lavalamp reviewed (commented):
Sorry for the delay. The interface seems OK now so I reviewed the implementation, too.

mandelsoft commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -569,6 +608,40 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+		return nil
+	}
+
+	if s.stopped {
```
Yes it does, if this is removed the `addCh` of the listener is closed twice, during the stop and the remove. This could be fixed by restructuring the existing code. But I would prefer to delegate this to a separate PR.

mandelsoft committed:
active remove/add tests for event handlers

atoato88 commented:
Gentle ping @mandelsoft
How about current state for this PR?

mandelsoft commented:
I added requested changes and rebases in October and added some comments to conversations to ask whether this is ok. But I didn't get answers up to now.

dims commented:
/assign @lavalamp

atoato88 commented:
@mandelsoft 
This PR needs rebase.
Could you do that?

atoato88 commented:
/remove-lifecycle stale

lavalamp commented:
/reopen

I think I was OOO when this last had activity and didn't see it

FillZpp commented:
👀  on this.

We have an issue https://github.com/kubernetes-sigs/controller-runtime/issues/1884 that proposes to remove watches dynamically, but controller-runtime couldn't do this unless client-go supports removal of event handlers.

alexzielenski commented:
@mandelsoft Do you have plans to continue contributing to this PR? If not I am happy to continue this work in another PR.

/assign alexzielenski

alexzielenski commented:
other pr is merged so closing this

/close
